### ISSUE TITLE
destroy the "Expect(err).NotTo(HaveOccurred())" anti-pattern in e2e tests

### ISSUE THREAD

**Issue by ixdy** [kubernetes/kubernetes#34059]
I'm getting tired of seeing flake issues like https://github.com/kubernetes/kubernetes/issues/33647 where every single failure is of the form

```
Expected error:
    <*errors.errorString | 0xc820412c10>: {
        s: "timed out waiting for the condition",
    }
    timed out waiting for the condition
not to have occurred
```

This is about as unhelpful an error message as you can get.

The problem is that we have

``` go
Expect(err).NotTo(HaveOccurred())
```

peppered throughout the code everywhere, and in this form, you can only figure out what failed if you read the Ginkgo error, remember which line has the line number you actually want, and then try to read surrounding code in that file.

Possible improvements given in http://onsi.github.io/gomega/#handling-errors:
- Use `Succeed()`
- Annotate the assertions
- Basically, do anything except just Expect(err).NotTo(HaveOccurred())

cc @kubernetes/sig-testing

**Comment by spxtr** [comment]
```
$ ag --noheading --nobreak "Expect\(err\)\.NotTo\(HaveOccurred\(\)\)" | wc -l
580
```
:cry:

**Comment by ixdy** [comment]
The github flake issues don't have that link. Maybe they should?

**Comment by ixdy** [comment]
Also, it's even worse:
```console
$ sift --no-group  'ExpectNoError\([^"]+\)' | wc -l
419
```

**Comment by spxtr** [comment]
```
$ ag --nobreak --noheading "Expect\(err\)\.NotTo\(HaveOccurred\(\)\)" | wc -l
603
$ ag --nobreak --noheading "Expect\(err\)\.NotTo\(HaveOccurred\(\), \".*\"\)" | wc -l
53
```
Think it's worth adding a presubmit verify check ensuring that the first number doesn't go up?

**Comment by jeremyKisner** [comment]
Is this believed to still be a problem?

**Comment by rmmh** [comment]
[Yes, it's still happening.](https://storage.googleapis.com/k8s-gubernator/triage/index.html?text=Expected%20error%3A(.%7C%5Cn)*not%20to%20have%20occurred)

**Comment by jeremyKisner** [comment]
Are the additional messages in the errors more helpful? I am not sure if removing the error message is a good approach. What do you think?

**Comment by spiffxp** [comment]
If the underlying err was more descriptive than "timed out waiting for the condition" this anti-pattern wouldn't be nearly as annoying.

**Comment by spiffxp** [comment]
/remove-lifecycle stale
this continues to be a great place for new contributors to help out

**Comment by louislepper** [comment]
I've raised a PR to improve the error messages in the DNS related e2e tests:

#67619

Just started learning Go/Kubernetes - all feedback welcome :-)

**Comment by nikhita** [comment]
@louislepper thanks for that PR!

I'm going to label this a good-first-issue because this involves fixing error messages.

If you are intersted to take this up, pick a few (related) e2e tests and follow the same methods as that of https://github.com/kubernetes/kubernetes/pull/67619 for improving the error messages. Please feel free to ask questions on slack if you get stuck.

/good-first-issue

**Comment by Pradip-Khakurel** [comment]
I've also raised a PR (#69993) to improve error messages in some e2e auth related tests. I tried to follow #67619.

I am new to Go/Kubernetes so all feedback are welcome.

**Comment by Zanadar** [comment]
I there a spot we could  add some sort of simple grep check that would run on git commits to prevent more of this sneaking in? I'd love to open a PR on that, I'm just not sure where it'd go.

**Comment by ixdy** [comment]
@Zanadar it might be some sort of `hack/verify-*.sh` check, perhaps.

**Comment by spiffxp** [comment]
/area deflake
More meaningful error text would help us track down flakes

**Comment by palash25** [comment]
Hi :wave: , I will be working on network/service.go file. Will send a PR soon. :smile:

**Comment by pohly** [comment]
/reopen

PR #72830 fixed some issues, but not all. PR #72728 is still open.

**Comment by oomichi** [comment]
We can use framework.ExpectNoError() instead of Expect(err).NotTo(HaveOccurred()) like https://github.com/kubernetes/kubernetes/pull/74113
That will lead we can avoid `. "github.com/onsi/gomega"` which is invalid against golint in long term.

**Comment by pohly** [comment]
But then `framework.ExpectNoError()` still needs an additional message string, doesn't it? Otherwise it's not better than `Expect(err).NotTo(HaveOccurred())`.

**Comment by alejandrox1** [comment]
This may be useful for sig-testing-commons xref: #75601 
/area e2e-test-framework

**Comment by sshukun** [comment]
I have raised a PR(#76692) trying to improve the error messages in bootstrap related file and I am also trying to follow #67619.
I am new to GO and kubernetes and this is my first PR, so I am happy to get any feedback.

**Comment by alejandrox1** [comment]
Some updates on this: the testing-commons project is working on refactoring a lot of the tests and the framework, which will deal with some of the unannotated tests, and we are also working on writing some e2e testing best practices to prevent more unannotated tests in the future.

**Comment by alejandrox1** [comment]
Leaving a question for the community here...
A lot of the work in testing commons has ended up being enforcing that tests spew out meaningful messages.
At this point in time, plain `Expect(err).NotTo(HaveOccurred())`  that output generic error messages on failure are not happening a whole lot and in many instances the error itself has plenty of information about the test.

Thus the issue raised here is largely resolved but there *may* be places where things could be improved.
Places where things could be improved (where an e2e tests requires additional annotation on error checking) are unkown and there is not a simple way of tracking them down (some instances of `Expect()` do not have an annotation on purpose because the possible error has already plenty of information).
Does anyone have any additional suggestions of work that should be done?

If not, then we can follow the lazy consensus route and just let fejta bot close this issue.

**Comment by neolit123** [comment]
> If not, then we can follow the lazy consensus route and just let fejta bot close this issue.
+1

**Comment by pohly** [comment]
> many instances the error itself has plenty of information about the test

I'd rather err on the side of caution and always add an explanation, because how can the test author be sure that *all* errors will be informative enough?

> Does anyone have any additional suggestions of work that should be done?

Can we document somewhere how to use `Expect(err).NotTo(HaveOccurred())` and `framework.ExpectNoError()` properly, i.e. "add an explanation unless it is certain that the error itself is informative"?

In the past, I've used this issue as reference in code reviews. It would be nice to have something a bit more concise and permanent.

**Comment by spiffxp** [comment]
About 50% of the Expect(No)Error calls still lack an explanation
```
spiffxp@spiffxp-macbookpro:e2e (master %)$ ag framework\\.ExpectNoError | grep -v , | wc -l
    1682
spiffxp@spiffxp-macbookpro:e2e (master %)$ ag framework\\.ExpectError | grep -v , | wc -l
      41
spiffxp@spiffxp-macbookpro:e2e (master %)$ ag framework\\.ExpectNoError | grep , | wc -l
    1630
spiffxp@spiffxp-macbookpro:e2e (master %)$ ag framework\\.ExpectError | grep , | wc -l
      38
```

However, the unexplained errs are happening far less frequently: https://storage.googleapis.com/k8s-gubernator/triage/index.html?text=not%20to%20have%20occurred

>Can we document somewhere how to use Expect(err).NotTo(HaveOccurred()) and framework.ExpectNoError() properly, i.e. "add an explanation unless it is certain that the error itself is informative"?

https://github.com/kubernetes/community/blob/fc892d4d1697f63794efa4b06e83d05092b470ff/contributors/devel/sig-testing/writing-good-e2e-tests.md#debuggability (which could probably use a rewrite, but I think this specific link addresses your concern)

**Comment by spiffxp** [comment]
/close
Since my comment just bumped the recency of this, I'm going to play the role of @fejta-bot for today.  

Rotten issues close after 30d of inactivity.
Reopen the issue with `/reopen`.
Mark the issue as fresh with `/remove-lifecycle rotten`.
